fish-rust/src/builtins/ulimit.rs

```diff
@@ -1,16 +1,468 @@
+use std::cmp::Ordering;
+
+use libc::{c_uint, rlim_t, RLIM_INFINITY};
+use nix::errno::Errno;
+use once_cell::sync::Lazy;
+
+use crate::{compat::*, fallback::*};
+
 use super::prelude::*;
 
+fn getrlimit(resource: c_uint) -> Result<(rlim_t, rlim_t), Errno> {
+    use nix::sys::resource;
+    let resource: i32 = resource.try_into().unwrap();
+    resource::getrlimit(unsafe { std::mem::transmute(resource) }) // Resource is #[repr(i32)] so this is ok
+}
+
+fn setrlimit(resource: c_uint, rlim_cur: rlim_t, rlim_max: rlim_t) -> Result<(), Errno> {
+    use nix::sys::resource;
+    let resource: i32 = resource.try_into().unwrap();
+    resource::setrlimit(
+        unsafe { std::mem::transmute(resource) }, // Resource is #[repr(i32)] so this is ok
+        rlim_cur,
+        rlim_max,
+    )
+}
+
+/// Print the value of the specified resource limit.
+fn print(resource: c_uint, hard: bool, streams: &mut IoStreams) {
+    let l = get(resource, hard);
+
+    if l == RLIM_INFINITY {
+        streams.out.append(wgettext_fmt!("unlimited\n"));
+    } else {
+        streams
+            .out
+            .append(wgettext_fmt!("%lu\n", l / get_multiplier(resource)));
+    }
+}
+
+/// Print values of all resource limits.
+fn print_all(hard: bool, streams: &mut IoStreams) {
+    let mut w = 0;
+
+    for resource in resource_arr.iter() {
+        w = w.max(fish_wcswidth(resource.desc));
+    }
+    for resource in resource_arr.iter() {
+        let (rlim_cur, rlim_max) = getrlimit(resource.resource).unwrap();
```

Comment:
I don't really think this approach is the best but made the change regardless. `set_limit` sets limits based on return values from `getrlimit` if the `soft`/`hard` flags are set. This means that if `getrlimit` fails, `setrlimit` can possibly be called with 0 instead of the intended/current value. Returning early before `setrlimit` (not necessarily crashing by unwrapping) would be best IMO.